apis/common/v1/connection_details.go

```diff
@@ -0,0 +1,157 @@
+/*
+Copyright 2019 The Crossplane Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1
+
+import "k8s.io/apimachinery/pkg/runtime"
+
+// PublishConnectionDetailsTo represents configuration of a connection secret.
+type PublishConnectionDetailsTo struct {
+	// Name is the name of the connection secret
+	Name string `json:"name"`
+
+	// Metadata is secret store specific key/value pairs to be used as metadata
+	// Please note, expected keys will differ for each store type. For example,
+	// it could be "labels" and "annotations" in case of "Kubernetes", but it
+	// would be "tags" for "AWS Secret Manager".
+	// +kubebuilder:pruning:PreserveUnknownFields
+	Metadata runtime.RawExtension `json:"metadata,omitempty"`
+
+	// SecretStoreConfigRef specifies which secret store config should be used
+	// for this ConnectionSecret.
+	// +kubebuilder:default={"name": "default"}
+	SecretStoreConfigRef *Reference `json:"configRef,omitempty"`
+}
+
+// SecretStoreType represents a secret store type.
+type SecretStoreType string
+
+const (
+	// SecretStoreKubernetes indicates that secret store type is
+	// Kubernetes. In other words, connection secrets will be stored as K8s
+	// Secrets.
+	SecretStoreKubernetes SecretStoreType = "Kubernetes"
+
+	// SecretStoreVault indicates that secret store type is Vault.
+	SecretStoreVault SecretStoreType = "Vault"
+)
+
+// SecretStoreConfig represents configuration of a Secret Store.
+type SecretStoreConfig struct {
+	// Type configures which secret store to be used. Only the configuration
+	// block for this store will be used and others will be ignored if provided.
+	// +kubebuilder:default=Kubernetes
+	Type SecretStoreType `json:"type,omitempty"`
+
+	// DefaultScope used for scoping secrets for "cluster-scoped" resources.
+	// If store type is "Kubernetes", this would mean the default namespace to
+	// store connection secrets for cluster scoped resources.
+	// In case of "Vault", this would be used as the default parent path.
+	// Typically, should be set as Crossplane installation namespace.
+	DefaultScope string `json:"defaultScope,omitempty"`
```

Comment:
Nit: Should this be `*string` and `// +optional`?

Here and a few other places in the type definitions (e.g. `Type`, above) I'm seeing `omitempty` tags on types that do not have nil zero values, and are missing `// +optional` tags. Please have a scan through and make sure anything that isn't optional has no `omitempty` tag, and anything that is optional has that tag, plus the optional marker and a nil zero value.